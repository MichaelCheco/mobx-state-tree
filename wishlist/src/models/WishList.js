import { types } from 'mobx-state-tree';

export const WishListItem = types
	.model({
		name: types.string,
		price: types.number,
		// image: types.optional(types.string, "")
		image: '',
	})
	.actions(self => {
		function changeName(newName) {
			self.name = newName;
		}
		return {
			changeName,
		};
	});

export const WishList = types.model({
	items: types.optional(types.array(WishListItem), []),
});
