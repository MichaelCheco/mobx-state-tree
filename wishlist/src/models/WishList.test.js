import { WishListItem, WishList } from './WishList';
import { getSnapshot, onSnapshot, onPatch } from 'mobx-state-tree';
it('can create an instance of a model', () => {
	const item = WishListItem.create({
		name: 'War and Peace',
		price: 32.99,
		// image: '',
	});
	expect(item.image).toBe('');
	expect(item.price).toBe(32.99);
	item.changeName('Fahrenheit 451');
	expect(item.name).toBe('Fahrenheit 451');
});

it('can create a wishlist', () => {
	const list = WishList.create({
		items: [
			{
				name: 'War and Peace',
				price: 32.99,
			},
		],
	});
	expect(list.items.length).toBe(1);
	expect(list.items[0].price).toBe(32.99);
});

it('can add new items', () => {
	const list = WishList.create();
	const states = [];
	onSnapshot(list, snapshot => {
		states.push(snapshot);
	});
	list.add({
		name: 'Musashi',
		price: 50.0,
	});
	expect(list.items.length).toBe(1);
	expect(list.items[0].price).toBe(50.0);
	list.items[0].changeName('Essentialism');
	expect(list.items[0].name).toBe('Essentialism');

	expect(getSnapshot(list)).toMatchSnapshot();
	expect(states).toMatchSnapshot();
	// expect(getSnapshot(list)).toEqual({
	// 	items: [
	// 		{
	// 			name: 'Essentialism',
	// 			price: 50.0,
	// 			image: '',
	// 		},
	// 	],
	// });
});
it('can add new items -2', () => {
	const list = WishList.create();
	const patches = [];
	onPatch(list, patch => {
		patches.push(patch);
	});
	list.add({
		name: 'Musashi',
		price: 50.0,
	});
	list.items[0].changeName('Essentialism');

	expect(patches).toMatchSnapshot();
});
