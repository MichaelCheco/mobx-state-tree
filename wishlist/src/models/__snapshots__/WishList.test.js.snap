// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can add new items -2 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "image": "",
      "name": "Musashi",
      "price": 50,
    },
  },
  Object {
    "op": "replace",
    "path": "/items/0/name",
    "value": "Essentialism",
  },
]
`;

exports[`can add new items 1`] = `
Object {
  "items": Array [
    Object {
      "image": "",
      "name": "Essentialism",
      "price": 50,
    },
  ],
}
`;

exports[`can add new items 2`] = `
Array [
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "Musashi",
        "price": 50,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "image": "",
        "name": "Essentialism",
        "price": 50,
      },
    ],
  },
]
`;
